### PR TITLE
GSLUX-559: Display external metadata

### PR DESCRIPTION
**Pull request by tkohr**
### JIRA issue

https://jira.camptocamp.com/browse/GSLUX-559

### Description

PR adds the capability to add WMTS layers to the remote layers layertree and to display their metadata.

The same should already be possible for WMS layers.

To-dos:

- [x] some more refactoring
- [x] adding some tests

Note: It does not address displaying the WMTS on the map.

### CODE REVIEW

File: src/components/layer-metadata/layer-metadata.vue [hunk 1]
```diff
@@ -61,10 +79,36 @@ function closeLayerMetadata() {
...
+            "
+            @click="hideFullDescription"
+            class="text-secondary hover:underline"
+          >
+            -
+          </button>
+        </span>
       </div>
       <layer-metadata-item
         v-if="layerMetadata.legalConstraints"
```
Comment (AlitaBernachot) [inline src/components/layer-metadata/layer-metadata.vue:114]: Missing css styling for `<h4>{{ t('Legend') }}</h4>`

Comment (AlitaBernachot) [reply src/components/layer-metadata/layer-metadata.vue:114]: Also the title 'avec légende' is visible even if there is not legend in production.

File: src/components/layer-metadata/layer-metadata.vue [hunk 2]
```diff
@@ -61,10 +79,36 @@ function closeLayerMetadata() {
...
+            "
+            v-if="
+              displayFullDescription &&
+              (description?.length || '') > MAX_DESCRIPTION_LENGTH
+            "
+            @click="hideFullDescription"
+            class="text-secondary hover:underline"
+          >
+            -
+          </button>
```
Comment (AlitaBernachot) [inline src/components/layer-metadata/layer-metadata.vue:110]: can this be done using only one button that would handle a toggle function or would it be too wordy?

**Review by AlitaBernachot — APPROVED** [review]
I have a few suggestions but really, it is up to you to make the changes or not.
Also, can you add just a few e2e tests please 😝?
Thanks for this great job!

File: src/components/remote-layers/remote-layers.vue [hunk 3]
```diff
@@ -24,9 +26,12 @@ const wmsLayers: ShallowRef<DropdownOptionModel[]> = shallowRef([])
...
 
 let isLoading = false
-let inputWmsUrl: string
-let currentWmsUrl: string
-let currentWmsEndpoint: OgcClientWmsEndpoint
+let inputRemoteUrl: string
+let currentRemoteUrl: string
+let currentRemoteEndpoint: {
+  type: 'WMS' | 'WMTS'
+  endpoint: OgcClientWmsEndpoint | WmtsEndpoint
```
[the remark this answers is not in the document]
Comment (AlitaBernachot) [reply src/components/remote-layers/remote-layers.vue]: Oops ignore this, I think you have done this in your last commits, can you confirm?

File: src/services/common/wmts.endpoint.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
+import WMTSCapabilities from 'ol/format/WMTSCapabilities.js'
+import {
+  WmtsServiceInfo,
+  WmtsCapabilitiesLayer,
+} from './../../components/remote-layers/remote-layers.model'
+
+export class WmtsEndpoint {
+  private _capabilitiesPromise: Promise<void>
```
[the remark this answers is not in the document]
Comment (tkohr) [reply src/services/common/wmts.endpoint.ts]: Yep, I agree completely a little too much inspiration from ogc-client ;-), got rid of the underscores

File: src/components/remote-layers/remote-layers.mapper.ts [hunk 5]
```diff
@@ -61,16 +90,30 @@ export function remoteLayerToLayer({
 }: {
   id: LayerId
   url: string
-  remoteLayer: OgcClientWmsLayerFull
+  remoteLayer: OgcClientWmsLayerFull | WmtsCapabilitiesLayer
```
[the remark this answers is not in the document]
Comment (tkohr) [reply src/components/remote-layers/remote-layers.mapper.ts]: I actually now removed most of the ogc-client types, hoping this was not too much

File: src/services/common/wmts.endpoint.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        return response.text()
+      })
+      .then(text => {
+        const result = parser.read(text)
+        this._serviceInfo = result.ServiceIdentification
+        this._layers = result.Contents?.Layer
+      })
+  }
+
+  isReady() {
```
[the remark this answers is not in the document]
Comment (tkohr) [reply src/services/common/wmts.endpoint.ts]: Left this as is. Where do you see this warning?

File: src/components/remote-layers/remote-layers.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (tkohr) [reply src/components/remote-layers/remote-layers.vue]: Yes, indeed :-)

File: src/components/remote-layers/remote-layers.vue [hunk 7]
```diff
@@ -46,48 +51,63 @@ remoteLayersService.fetchRemoteWmsEndpoint().then(wmsLayersFetch => {
   }))
 })
 
-async function getWmsEndpoint(url: string) {
+async function getRemoteEndpoint(url: string) {
```
[the remark this answers is not in the document]
Comment (tkohr) [reply src/components/remote-layers/remote-layers.vue:50]: Moved it to the service

File: src/services/common/wmts.helper.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    const layer = wmtsEndpoint?.getLayerByName(idValues.layerName)
+    return {
+      title: layer.Title,
+      description: layer.Abstract,
+      accessConstraints: service.AccessConstraints,
+      serviceDescription: service.Abstract,
+    }
+  }
+}
+export const wmtsHelper = new WmtsHelper()
```
[the remark this answers is not in the document]
Comment (tkohr) [reply src/services/common/wmts.helper.ts]: I added an abstract `OwsHelper` that is extended by the `WmtsHelper`

File: src/services/layer-metadata/layer-metadata.service.ts [hunk 9]
```diff
@@ -74,8 +75,7 @@ export class LayerMetadataService {
       if (idValues.serviceType === 'WMS') {
         metadata = await wmsHelper.getMetadata(idValues)
       } else if (idValues.serviceType == 'WMTS') {
-        // TODO: handle WMTS
-        // metadata = appWmtsHelper.getMetadata(metadataUid)
+        metadata = await wmtsHelper.getMetadata(idValues)
```
[the remark this answers is not in the document]
Comment (tkohr) [reply src/services/layer-metadata/layer-metadata.service.ts]: I added an abstract `OwsHelper` and a `RemoteMetadataHelper` as a "proxy" hoping this makes this cleaner without adding too much for one method (so far)

File: src/services/layer-metadata/layer-metadata.service.ts [hunk 10]
```diff
@@ -74,8 +75,7 @@ export class LayerMetadataService {
       if (idValues.serviceType === 'WMS') {
         metadata = await wmsHelper.getMetadata(idValues)
       } else if (idValues.serviceType == 'WMTS') {
```
[the remark this answers is not in the document]
Comment (tkohr) [reply src/services/layer-metadata/layer-metadata.service.ts]: Added an enum in the model

File: src/services/layer-metadata/layer-metadata.service.spec.ts [hunk 11]
```diff
@@ -132,4 +133,23 @@ describe('LayerMetadataService', () => {
       expect(metadata).toEqual(metadataMock)
     })
   })
+  describe('#getLayerMetadata with WMTS layer', () => {
+    let spyWMTSMetadata
```
[the remark this answers is not in the document]
Comment (tkohr) [reply src/services/layer-metadata/layer-metadata.service.spec.ts]: Also the spies ?

File: src/services/remote-layers/remote-layers.mapper.ts [hunk 12]
```diff
@@ -4,12 +4,12 @@ import { LayerTreeNodeModel } from '@/components/layer-tree/layer-tree.model'
 import { RemoteLayer } from './remote-layers.model'
 import { remoteLayersService } from './remote-layers.service'
 
-function sortLayerTreeNoChildrenFirst(a: RemoteLayer, b: RemoteLayer) {
-  if (a.children && !b.children) {
+export function sortLayerTreeNoChildrenFirst(a: RemoteLayer, b: RemoteLayer) {
```
[the remark this answers is not in the document]
Comment (tkohr) [reply src/services/remote-layers/remote-layers.mapper.ts]: Adapted the test suit accordingly

File: src/services/layer-metadata/layer-metadata.service.ts [hunk 13]
```diff
@@ -19,33 +19,26 @@ export class LayerMetadataService {
   // TODO: get from config or relative
   private localMetadataBaseUrl = 'https://map.geoportail.lu/getMetadata'
 
-  async getLayerMetadata(
-    id: LayerId,
-    currentLanguage: string
-  ): Promise<LayerMetadataModel> {
+  async getLayerMetadata(id: LayerId, currentLanguage: string) {
     const layer: ThemeNodeModel | undefined =
       useThemes().findById(+id) || useThemes().findBgLayerById(+id)
```
[the remark this answers is not in the document]
Comment (tkohr) [reply src/services/layer-metadata/layer-metadata.service.ts]: Good catch, I inversed the order

File: src/components/layer-metadata/layer-metadata.vue (hunk 1 above)
Comment (tkohr) [reply src/components/layer-metadata/layer-metadata.vue:114]: Thought it'd be more logic to not display "avec légende" if there is none. Didn't like the style on production, but you're right I adapted the font size accordingly.

File: src/components/layer-metadata/layer-metadata.vue (hunk 2 above)
Comment (tkohr) [reply src/components/layer-metadata/layer-metadata.vue:110]: Kept it as is for now, as the content and the title would become conditional.

**Comment by tkohr** [comment]
Thanks a lot for the detailed review @AlitaBernachot! Don't hesitate to have another look at my latest commit, as it includes quite some changes :-)

File: src/services/common/wmts.endpoint.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (AlitaBernachot) [reply src/services/common/wmts.endpoint.ts]: I can't remember where I saw this warning... 😅

File: src/services/common/wmts.helper.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (AlitaBernachot) [reply src/services/common/wmts.helper.ts]: Sorry, I meant `static` for `getMetadata()` method, because I don't see why this class must be instantiate.
But this is not very a big deal. Ok for extending the abstract as well.

File: src/components/layer-metadata/layer-metadata.vue (hunk 1 above)
Comment (AlitaBernachot) [reply src/components/layer-metadata/layer-metadata.vue:114]: I also do not like the style in production. + agree with you, it is more logic to hide the title when there is no legend.

File: src/services/layer-metadata/remote-metadata.helper.ts [hunk 14]
```diff
@@ -0,0 +1,28 @@
+import { REMOTE_SERVICE_TYPE } from '../remote-layers/remote-layers.model'
+import { WmsHelper } from './wms.helper'
+import { WmtsHelper } from './wmts.helper'
+
+export class RemoteMetadataHelper {
+  private wmsHelper: WmsHelper
+  private wmtsHelper: WmtsHelper
```
Comment (AlitaBernachot) [inline src/services/layer-metadata/remote-metadata.helper.ts:7]: is there a reason for instantiating the helpers here? and store them as properties in the current helper?
Maybe I am missing something but you could simply return wmsHelper.getMetadata or wmtsHelper.getMetadata directly.

**Review by AlitaBernachot — APPROVED** [review]
Thanks for the update! 👍

File: src/services/layer-metadata/remote-metadata.helper.ts (hunk 14 above)
Comment (tkohr) [reply src/services/layer-metadata/remote-metadata.helper.ts]: Of course, thanks for paying attention to this, as I did not :-)
